File: C++/BinaryTree/isSameTree.cpp
/*
Solution: 递归，时间复杂度O(n)
- 说明：代码参考高票答案，简洁规范；注释掉的代码是自己写的代码，看起来有点混乱...
*/
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        if(p == NULL || q == NULL)
            return p == q;
        return(p->val == q->val && isSameTree(p->left, q->left) && isSameTree(p->right, q->right));
 
//         if(!p && !q)
//             return true;
//         if((p && !q) || (!p && q) || (p->val != q->val))
//             return false;
//         else
//             return isSameTree(p->left, q->left) && isSameTree(p->right, q->right);
    }
};

File: Python3/BinaryTree/isSameTree.py
# -*- coding:UTF-8 -*-
'''
Solution:
- 注意使用 p == q无法判断这两个节点是否相同，即使他们val相同，left，right都相同，使用p==q还是判为FALSE；
- 要用.val来判断，再分别对其左右的val判断
'''

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isSameTree(self, p, q):
        """
        :type p: TreeNode
        :type q: TreeNode
        :rtype: bool
        """
        if not(p and q):   # 说明p,q中至少有一个为None
            return p == q
        
        if p.val == q.val:
            return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)
    
        return False
        